apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1/UrlSigner.V4Signer.cs

```diff
@@ -156,6 +176,78 @@ internal SigningState(RequestTemplate template, Options options, IBlobSigner blo
                     $"{_scheme}://{_host}{_resourcePath}?{_canonicalQueryString}&X-Goog-Signature={WebUtility.UrlEncode(signature)}";
             }
 
+            /// <summary>
+            /// State which needs to be carried between the "pre-signing" stage and "post-signing" stages
+            /// of the implementation.
+            /// </summary>
+            private readonly struct PostPolicySigningState
+            {
+                private readonly PostPolicy _policy;
+                private readonly string _encodedPolicy;
+                internal readonly byte[] _blobToSign;
+                internal readonly DateTimeOffset _expiration;
+                private readonly Uri _url;
+
+                internal PostPolicySigningState(PostPolicy policy, Options options, IBlobSigner blobSigner, IClock clock)
+                {
+                    string host = options.UrlStyle switch
+                    {
+                        UrlStyle.PathStyle => policy.Bucket == null ? StorageHost : $"{StorageHost}/{policy.Bucket}",
+                        UrlStyle.VirtualHostedStyle => policy.Bucket == null ?
+                            throw new ArgumentNullException(nameof(PostPolicy.Bucket), $"When using {UrlStyle.VirtualHostedStyle} a bucket condition must be set in the policy.") :
+                            $"{policy.Bucket}.{StorageHost}",
+                        UrlStyle.BucketBoundHostname => options.BucketBoundHostname,
+                        _ => throw new ArgumentOutOfRangeException(nameof(options.UrlStyle))
+                    };
+
+                    options = options.ToExpiration(clock);
+                    var now = clock.GetCurrentDateTimeUtc();
+
+                    int expirySeconds = (int)(options.Expiration.Value - now).TotalSeconds;
+                    if (expirySeconds <= 0)
+                    {
+                        throw new ArgumentOutOfRangeException(nameof(options.Expiration), "Expiration must be at least 1 second");
+                    }
+                    if (expirySeconds > MaxExpirySecondsInclusive)
+                    {
+                        throw new ArgumentOutOfRangeException(nameof(options.Expiration), "Expiration must not be greater than 7 days.");
+                    }
+
+                    var datestamp = now.ToString("yyyyMMdd", CultureInfo.InvariantCulture);
+                    string credentialScope = $"{datestamp}/{DefaultRegion}/{ScopeSuffix}";
+
+                    policy.SetField(PolicyCreationDateTime.Element, now);
+                    policy.SetField(PolicyAlgorithm.Element, Algorithm);
+                    policy.SetField(PolicyCredential.Element, $"{blobSigner.Id}/{credentialScope}");
```

Comment:
I'm confused with fields such as `key`, `acl` are added into the signature in this case, but should be caught with conformance tests if there's an issue.